### PR TITLE
add pre/post run functions to threadpool scheduler.

### PR DESCRIPTION
**Pull request by amosbird**
https://github.com/Amanieu/asyncplusplus/issues/23

### CODE REVIEW

File: src/threadpool_scheduler.cpp [hunk 1]
```diff
@@ -172,6 +180,10 @@ static void thread_task_loop(threadpool_data* impl, std::size_t thread_id, task_
 	// Flag indicating if we have added a continuation to the task
 	bool added_continuation = false;
 
+    if (impl->prerun) impl->prerun();
+
+    SCOPE_EXIT({if (impl->postrun) impl->postrun();});
+
```
Comment (Amanieu) [inline src/threadpool_scheduler.cpp]: This should be done in `worker_thread`, not here.

Also no exceptions are thrown by this code, so you don't need all the complicated scoping stuff: just call `prerun` before the call to `thread_task_loop` and `postrun` after.

File: src/internal.h [hunk 2]
```diff
@@ -86,6 +86,30 @@
...
+template <class F>
+inline scope_guard<F> make_scope_guard(F && function) { return std::forward<F>(function); }
+
+}
+
+#define SCOPE_EXIT_CONCAT(n, ...) \
+const auto scope_exit##n = async::make_scope_guard([&] { __VA_ARGS__; })
+#define SCOPE_EXIT_FWD(n, ...) SCOPE_EXIT_CONCAT(n, __VA_ARGS__)
+#define SCOPE_EXIT(...) SCOPE_EXIT_FWD(__LINE__, __VA_ARGS__)
+
```
Comment (Amanieu) [inline src/internal.h]: There's no need for all this stuff. See my other comment.

File: src/threadpool_scheduler.cpp (hunk 1 above)
Comment (amosbird) [reply src/threadpool_scheduler.cpp]: Hmm, but there are two early return calls inside the loop.

Comment (Amanieu) [reply src/threadpool_scheduler.cpp]: Which is why I said to put the call to `prerun` and `postun` in the `worker_thread` function instead.

**Pull request merged**